File: backend/src/schemas/health.py
"""Health check response schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class HealthResponse(BaseModel):
    """Response model for health check endpoint."""

    status: Literal["healthy", "degraded", "unhealthy"] = Field(
        description="Current health status of the API"
    )
    version: str = Field(description="API version string")
    timestamp: datetime = Field(description="Current server timestamp (UTC)")

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "status": "healthy",
                    "version": "1.0.0",
                    "timestamp": "2024-12-02T10:30:00Z",
                }
            ]
        }
    }


class ModelInfo(BaseModel):
    """Information about a single ML model."""

    name: str = Field(description="Model name/identifier")
    loaded: bool = Field(description="Whether the model is loaded in memory")
    type: str = Field(description="Model type (e.g., xgboost, linear_regression)")


class ModelsStatusResponse(BaseModel):
    """Response model for models status endpoint."""

    total: int = Field(description="Total number of models")
    ready: int = Field(description="Number of models loaded and ready")
    models: list[ModelInfo] = Field(description="Details for each model")
    timestamp: datetime = Field(description="Current server timestamp (UTC)")

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "total": 3,
                    "ready": 3,
                    "models": [
                        {"name": "xgboost", "loaded": True, "type": "xgboost"},
                        {"name": "linear_regression", "loaded": True, "type": "sklearn"},
                        {"name": "decision_tree", "loaded": True, "type": "sklearn"},
                    ],
                    "timestamp": "2024-12-02T10:30:00Z",
                }
            ]
        }
    }

File: packages/shared/src/schemas/health.ts
/**
 * Health check response schemas.
 * Source: backend/src/schemas/health.py
 */
import { z } from 'zod';

export const HealthStatus = z.enum(['healthy', 'degraded', 'unhealthy']);

/**
 * Response model for health check endpoint.
 * Source: backend/src/schemas/health.py::HealthResponse
 */
export const HealthResponseSchema = z.object({
  status: HealthStatus.describe('Current health status of the API'),
  version: z.string().describe('API version string'),
  timestamp: z.string().datetime().describe('Current server timestamp (UTC)'),
});

/**
 * Information about a single ML model.
 * Source: backend/src/schemas/health.py::ModelInfo
 */
export const ModelInfoSchema = z.object({
  name: z.string().describe('Model name/identifier'),
  loaded: z.boolean().describe('Whether the model is loaded in memory'),
  type: z.string().describe('Model type (e.g., xgboost, linear_regression)'),
});

/**
 * Response model for models status endpoint.
 * Source: backend/src/schemas/health.py::ModelsStatusResponse
 */
export const ModelsStatusResponseSchema = z.object({
  total: z.number().int().describe('Total number of models'),
  ready: z.number().int().describe('Number of models loaded and ready'),
  models: z.array(ModelInfoSchema).describe('Details for each model'),
  timestamp: z.string().datetime().describe('Current server timestamp (UTC)'),
});

export type HealthResponse = z.infer<typeof HealthResponseSchema>;
export type HealthStatusType = z.infer<typeof HealthStatus>;
export type ModelInfo = z.infer<typeof ModelInfoSchema>;
export type ModelsStatusResponse = z.infer<typeof ModelsStatusResponseSchema>;

